core/src/replay_stage.rs

```diff
@@ -3111,6 +3206,18 @@ impl ReplayStage {
 
             assert_eq!(bank_slot, bank.slot());
             if bank.is_complete() {
+                if let Some(first_alpenglow_slot) = first_alpenglow_slot {
+                    if !*is_alpenglow_migration_complete && bank.slot() >= first_alpenglow_slot {
+                        info!(
+                            "initiating alpenglow migration from replaying bank {}",
+                            bank.slot()
+                        );
+                        Self::initiate_alpenglow_migration(
+                            poh_recorder,
+                            is_alpenglow_migration_complete,
+                        );
+                    }
+                }
```

Comment:
Just making sure I have this straight:
- Higher up, we detect when someone has slated AG for activation from a root bank
- Here, we initiate AG migration once we have a complete Bank in the new epoch

Does it matter if the Bank that flips the switch ends up being rooted or not ? Thinking of current situation where we have some forking around epoch boundaries

Reply:
That's correct. It doesn't matter if the first bank we observe in the new epoch ends up being rooted. 

Also for the most part this migration logic is a placeholder - the critical section is the enabled from genesis case. 

The actual migration is still under debate and will likely involve a process where we try to switch to alpenglow at some slot - if the migration "fails" (details on this tbd) then we switch back to poh and try again later.